File: universal_printer_flutter/android/src/main/kotlin/com/universalprinter/universal_printer_flutter/Bridge.kt
package com.universalprinter.universal_printer_flutter

import android.graphics.BitmapFactory
import com.universalprinter.model.Align
import com.universalprinter.model.BarcodeSymbology
import com.universalprinter.model.Column
import com.universalprinter.model.CutType
import com.universalprinter.model.PaperWidth
import com.universalprinter.model.PrintDocument
import com.universalprinter.model.PrinterMessages
import com.universalprinter.model.PrintResult
import com.universalprinter.model.PrintType
import com.universalprinter.model.QrErrorLevel
import com.universalprinter.model.RenderMode
import com.universalprinter.model.TextSize
import com.universalprinter.model.printDocument
import com.universalprintersearch.model.DiscoveredPrinter

/**
 * JSON<->Kotlin translation for the MethodChannel contract. Kept separate from the plugin wiring so
 * the serialization shape is easy to audit against the Dart models. All enum values cross the channel
 * as their Kotlin `.name` string; unknown/absent values fall back to the SDK's own defaults.
 */
internal object Bridge {

    // ---- Discovery output ----

    fun printerToMap(p: DiscoveredPrinter): Map<String, Any?> = mapOf(
        "name" to p.name,
        "connectionType" to p.connectionType.name,
        "ipAddress" to p.ipAddress,
        "port" to p.port,
        "macAddress" to p.macAddress,
        "serialNumber" to p.serialNumber,
        "brand" to p.brand.name,
        "model" to p.model,
        "emulation" to p.emulation,
        "vendorId" to p.vendorId,
        "productId" to p.productId,
        "usbDeviceName" to p.usbDeviceName,
        "isImpact" to p.isImpact,
        "isBuiltIn" to p.isBuiltIn,
        "supportedPaperWidthsMm" to p.supportedPaperWidthsMm,
        "effectiveEmulation" to p.effectiveEmulation,
    )

    fun printersToList(list: List<DiscoveredPrinter>): List<Map<String, Any?>> = list.map(::printerToMap)

    // ---- Print result output ----

    fun resultToMap(r: PrintResult): Map<String, Any?> = when (r) {
        is PrintResult.Success -> mapOf(
            "status" to "success",
            "warnings" to r.warnings.map { it.name },
            "warningMessages" to r.warnings.map { PrinterMessages.warningMessage(it) },
        )
        is PrintResult.Error -> mapOf(
            "status" to "error",
            "reason" to r.reason.name,
            "userMessage" to PrinterMessages.userMessage(r.reason), // always safe to show the operator
            "details" to r.message,                                 // technical detail — for logging/support
            "message" to r.message,                                 // back-compat (== details)
        )
    }

    // ---- PrintDocument input (Dart JSON -> Kotlin) ----

    fun buildDocument(map: Map<*, *>): PrintDocument {
        val paper = enumOrDefault(map["paper"] as? String, PaperWidth.MM_80)
        return printDocument(paper) {
            (map["cut"] as? String)?.let { cut = enumOrDefault(it, CutType.PARTIAL) }
            (map["openDrawer"] as? Boolean)?.let { openDrawer = it }
            (map["renderMode"] as? String)?.let { renderMode = enumOrDefault(it, RenderMode.AUTO) }

            val elements = map["elements"] as? List<*> ?: emptyList<Any?>()
            for (raw in elements) {
                val e = raw as? Map<*, *> ?: continue
                when (e["type"] as? String) {
                    "text" -> text(
                        text = e["text"] as? String ?: "",
                        align = align(e["align"]),
                        bold = bool(e["bold"]),
                        underline = bool(e["underline"]),
                        invert = bool(e["invert"]),
                        size = enumOrDefault(e["size"] as? String, TextSize.NORMAL),
                    )
                    "columns" -> {
                        val cells = (e["cells"] as? List<*>).orEmpty().mapNotNull { c ->
                            (c as? Map<*, *>)?.let {
                                Column(it["text"] as? String ?: "", intOr(it["weight"], 1), align(it["align"]))
                            }
                        }
                        columns(*cells.toTypedArray())
                    }
                    "row" -> row(
                        left = e["left"] as? String ?: "",
                        right = e["right"] as? String ?: "",
                        leftWeight = intOr(e["leftWeight"], 1),
                        rightWeight = intOr(e["rightWeight"], 1),
                    )
                    "imageUrl" -> imageUrl(
                        url = e["url"] as? String ?: "",
                        align = align(e["align"]),
                        invert = bool(e["invert"]),
                        dither = bool(e["dither"]),
                    )
                    "image" -> {
                        val bytes = e["bytes"] as? ByteArray
                        val bmp = bytes?.let { BitmapFactory.decodeByteArray(it, 0, it.size) }
                        if (bmp != null) image(bmp, align(e["align"]), bool(e["invert"]), bool(e["dither"]))
                    }
                    "barcode" -> barcode(
                        data = e["data"] as? String ?: "",
                        symbology = enumOrDefault(e["symbology"] as? String, BarcodeSymbology.CODE128),
                        heightDots = intOr(e["heightDots"], 100),
                        align = align(e["align"]),
                    )
                    "qr" -> qr(
                        data = e["data"] as? String ?: "",
                        sizeDots = intOr(e["sizeDots"], 200),
                        errorLevel = enumOrDefault(e["errorLevel"] as? String, QrErrorLevel.M),
                        align = align(e["align"]),
                    )
                    "feed" -> feed(intOr(e["lines"], 1))
                    "divider" -> divider()
                    "raw" -> (e["bytes"] as? ByteArray)?.let { raw(it) }
                }
            }
        }
    }

    // ---- helpers ----

    private fun align(v: Any?): Align = enumOrDefault(v as? String, Align.LEFT)
    private fun bool(v: Any?): Boolean = v as? Boolean ?: false
    private fun intOr(v: Any?, default: Int): Int = when (v) {
        is Int -> v
        is Long -> v.toInt()
        is Number -> v.toInt()
        else -> default
    }

    private inline fun <reified T : Enum<T>> enumOrDefault(name: String?, default: T): T =
        name?.let { runCatching { enumValueOf<T>(it.uppercase()) }.getOrNull() } ?: default
}

File: universal_printer_flutter/ios/universal_printer_flutter/Sources/universal_printer_flutter/Bridge.swift
import Foundation
import Flutter

// Wire <-> Swift translation, mirror of Android's Bridge.kt. Parses the `document` map into a
// ReceiptDocument (same "type" switch) and builds the PrintResult dicts.
enum Bridge {

    static func buildDocument(_ map: [String: Any]) -> ReceiptDocument {
        let paper = PaperWidth.from(map["paper"] as? String)
        let cut = CutType.from(map["cut"] as? String)
        var elements: [ReceiptElement] = []

        for item in (map["elements"] as? [Any]) ?? [] {
            guard let e = item as? [String: Any] else { continue }
            switch (e["type"] as? String) ?? "" {
            case "text":
                elements.append(.text(str(e["text"]), align: TextAlign.from(e["align"] as? String),
                                      bold: bool(e["bold"]), underline: bool(e["underline"]),
                                      invert: bool(e["invert"]), size: TextSize.from(e["size"] as? String)))
            case "columns":
                let cells = ((e["cells"] as? [Any]) ?? []).compactMap { c -> ReceiptColumn? in
                    guard let cm = c as? [String: Any] else { return nil }
                    return ReceiptColumn(text: str(cm["text"]), weight: intOr(cm["weight"], 1),
                                         align: TextAlign.from(cm["align"] as? String))
                }
                elements.append(.columns(cells))
            case "row":
                elements.append(.columns([
                    ReceiptColumn(text: str(e["left"]), weight: intOr(e["leftWeight"], 1), align: .left),
                    ReceiptColumn(text: str(e["right"]), weight: intOr(e["rightWeight"], 1), align: .right),
                ]))
            case "imageUrl":
                elements.append(.imageUrl(str(e["url"]), align: TextAlign.from(e["align"] as? String),
                                          invert: bool(e["invert"]), dither: bool(e["dither"])))
            case "image":
                if let d = bytes(e["bytes"]) {
                    elements.append(.image(d, align: TextAlign.from(e["align"] as? String),
                                           invert: bool(e["invert"]), dither: bool(e["dither"])))
                }
            case "barcode":
                elements.append(.barcode(str(e["data"]), symbology: (e["symbology"] as? String) ?? "CODE128",
                                         heightDots: intOr(e["heightDots"], 100),
                                         align: TextAlign.from(e["align"] as? String)))
            case "qr":
                elements.append(.qr(str(e["data"]), sizeDots: intOr(e["sizeDots"], 200),
                                    errorLevel: (e["errorLevel"] as? String) ?? "M",
                                    align: TextAlign.from(e["align"] as? String)))
            case "feed":
                elements.append(.feed(intOr(e["lines"], 1)))
            case "divider":
                elements.append(.divider)
            case "raw":
                if let d = bytes(e["bytes"]) { elements.append(.raw(d)) }
            default:
                break
            }
        }
        return ReceiptDocument(paper: paper, cut: cut, elements: elements)
    }

    static func successResult(_ warnings: [String] = []) -> [String: Any] {
        ["status": "success", "warnings": warnings,
         "warningMessages": warnings.map { PrinterMessages.warningMessage($0) }]
    }
    /// [reason] is a PrintErrorReason wire value; [details] is the technical string (logged, not shown).
    static func errorResult(_ reason: String, _ details: String?) -> [String: Any] {
        ["status": "error", "reason": reason,
         "userMessage": PrinterMessages.userMessage(reason), // safe to show the operator
         "details": details ?? "",
         "message": details ?? ""]                            // back-compat (== details)
    }

    // MARK: value helpers (NSNull-safe)

    private static func str(_ v: Any?) -> String { (v as? String) ?? "" }
    private static func bool(_ v: Any?) -> Bool {
        if let b = v as? Bool { return b }
        if let n = v as? NSNumber { return n.boolValue }
        return false
    }
    private static func intOr(_ v: Any?, _ d: Int) -> Int {
        if let i = v as? Int { return i }
        if let n = v as? NSNumber { return n.intValue }
        return d
    }
    private static func bytes(_ v: Any?) -> Data? {
        if let t = v as? FlutterStandardTypedData { return t.data }
        if let d = v as? Data { return d }
        return nil
    }
}
